### PR TITLE
Revisited endpoint state restore

### LINKED ISSUE #1140: Verify endpoint state restore
- [x] Verify that restarting Cilium pod/ds correctly restores state of previous endpoints/pods
- [x] Add test to k8s CI test

### PR DESCRIPTION
**Pull request by aanm**
Fixes #1140

### CODE REVIEW

File: daemon/state.go [hunk 1]
```diff
@@ -219,8 +226,8 @@ func (d *Daemon) syncLabels(ep *endpoint.Endpoint) error {
 	return nil
 }
 
-// cleanUpDockerDandlingEndpoints cleans all endpoints that are dandling by checking out
-// if a particular endpoint has its container running.
+// cleanUpDockerDandlingEndpoints cleans all endpoints that are dandling by
```
Comment (ianvernon) [inline daemon/state.go]: Nit: Dandling --> Dangling

Comment (aanm) [reply daemon/state.go]: Fixed!

File: daemon/state.go [hunk 2]
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
...
-		}
-
 		endpointmanager.Insert(ep)
 		restored++
 
-		log.Infof("Restored endpoint: %d", ep.ID)
+		ep.Mutex.Unlock()
 	}
 
 	log.Infof("Restored %d endpoints", restored)
```
Comment (tgraf) [inline daemon/state.go]: This number of restored endpoints seems premature as it may fail later. I suggest we remove it.

File: daemon/state.go [hunk 3]
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
...
 	for k := range endpointmanager.Endpoints {
 		ep := endpointmanager.Endpoints[k]
-		if err := d.allocateIPs(ep); err != nil {
-			log.Errorf("Failed while reallocating ep %d's IP addresses: %s. Endpoint won't be restored", ep.ID, err)
-			d.deleteEndpoint(ep)
-			continue
-		}
+		go func(ep *endpoint.Endpoint, wg *sync.WaitGroup) {
+			if err := d.allocateIPs(ep); err != nil {
+				log.Errorf("Failed while reallocating ep %d's IP addresses: %s. Endpoint won't be restored", ep.ID, err)
```
Comment (tgraf) [inline daemon/state.go]: Let's be consistent and call it endpoint in all info messages
`"Failed to re-allocate IP of endpoint %d. Not restoring endpoint."`

File: daemon/state.go [hunk 4]
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
...
-		epID := ep.ID
-		if ep.SecLabel != nil {
-			epLabels := ep.SecLabel.DeepCopy()
-			ep.Mutex.RUnlock()
-			d.events <- *events.NewEvent(events.IdentityAdd, epLabels)
-		} else {
-			ep.Mutex.RUnlock()
-		}
-		log.Infof("Restored endpoint %d", epID)
+			log.Infof("EP %d's IP addresses successfully reallocated", ep.ID)
```
Comment (tgraf) [inline daemon/state.go]: This info message seems useless as it again may fail afterwards.

File: daemon/state.go [hunk 5]
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
...
+			ep.Mutex.RLock()
+			epID := ep.ID
+			if ep.SecLabel != nil {
+				epLabels := ep.SecLabel.DeepCopy()
+				ep.Mutex.RUnlock()
+				d.events <- *events.NewEvent(events.IdentityAdd, epLabels)
+			} else {
+				ep.Mutex.RUnlock()
+			}
+			log.Infof("Restored endpoint %d", epID)
```
Comment (tgraf) [inline daemon/state.go]: This is the most important info message, we may want to report the IP addresses as well.

File: daemon/state.go [hunk 6]
```diff
@@ -47,67 +48,73 @@ func (d *Daemon) SyncState(dir string, clean bool) error {
...
+			wg.Done()
+		}(ep, &wg)
 	}
 	endpointmanager.Mutex.Unlock()
 
+	wg.Wait()
+
```
Comment (tgraf) [inline daemon/state.go]: `"Restored %d endpoints"`

File: daemon/docker_watcher.go [hunk 7]
```diff
@@ -89,8 +89,8 @@ func (d *Daemon) SyncDocker() {
 
 func (d *Daemon) backgroundContainerSync() {
 	for {
-		d.SyncDocker()
 		time.Sleep(syncRateDocker)
+		d.SyncDocker()
```
Comment (tgraf) [inline daemon/docker_watcher.go:93]: Typo in commit title. "Spawning"

**Review by tgraf — CHANGES_REQUESTED** [review]

File: daemon/state.go (hunk 6 above)
Comment (aanm) [reply daemon/state.go]: Change it a little bit, better?

```
"Restored endpoint 38939 with IPs 10.15.250.189 and f00d::a00:20f:0:981b"
"Found 3 endpoints on which 3 were restored"
```

Comment (tgraf) [reply daemon/state.go]: nit: `.. of which 3 were restored`

Comment (tgraf) [reply daemon/state.go]: Otherwise looks great now!

**Review by tgraf — CHANGES_REQUESTED** [review]
Only a small nit left

File: daemon/state.go (hunk 6 above)
Comment (aanm) [reply daemon/state.go]: Fixed

**Comment by aanm** [comment]
@tgraf the tests (k8s, docker and unit tests) have passed. I've changed the flag restore functionality to be `true` as default.

**Review by tgraf — APPROVED** [review]

**Pull request merged**